Support sparse.sum on empty sparse tensor

pearu opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/65394

The masked sum on a full input tensor (of any layout) with an all-true mask is the same as the sum on the strided input tensor (after applying `to_dense` to sparse inputs). 
Since masked sum uses `torch.sparse.sum` then, for the simplicity of masked reductions implementations, its reduction behavior ought to be defined by the behavior of the `torch.sum`. This PR implements the behavioral connection with respect to the directional summation of empty sparse tensors that correspond to all-zero strided tensors.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #71239
* __->__ #71091
* #71073



cc @nikitaved @pearu @cpuhrsch

Differential Revision: [D33651750](https://our.internmc.facebook.com/intern/diff/D33651750)

pearu committed:
Support sparse.sum on empty sparse tensor

[ghstack-poisoned]

cpuhrsch commented:
Could you also please add a paragraph to the PR description that briefly describes why this is needed? I see you linked this to 65394, but it's not immediately obvious to me why this is needed or how it resolves the issue.

pearu commented:
> Could you also please add a paragraph to the PR description that briefly describes why this is needed? I see you linked this to 65394, but it's not immediately obvious to me why this is needed or how it resolves the issue.

@cpuhrsch sure. In short, it is a small detail of the implementation of masked sum on COO tensors, see the workaround comments in
https://github.com/pytorch/pytorch/pull/64809/files#diff-6d0bd55f94dcbd4ccf83f4006976fa4cb87eb06b72dc16238a21980226be6315R334-R392

pearu committed:
Update on "Support sparse.sum on empty sparse tensor"

Fixes https://github.com/pytorch/pytorch/issues/65394




cc nikitaved pearu cpuhrsch

[ghstack-poisoned]

pearu committed:
Update on "Support sparse.sum on empty sparse tensor"

Fixes https://github.com/pytorch/pytorch/issues/65394

The masked sum on a full input tensor (of any layout) with an all-true mask is the same as the sum on the strided input tensor (after applying `to_dense` to sparse inputs). 
Since masked sum uses `torch.sparse.sum` then, for the simplicity of masked reductions implementations, its reduction behavior ought to be defined by the behavior of the `torch.sum`. This PR implements the behavioral connection with respect to the directional summation of empty sparse tensors that correspond to all-zero strided tensors.




cc nikitaved pearu cpuhrsch

[ghstack-poisoned]

pearu committed:
Update on "Support sparse.sum on empty sparse tensor"

Fixes https://github.com/pytorch/pytorch/issues/65394

The masked sum on a full input tensor (of any layout) with an all-true mask is the same as the sum on the strided input tensor (after applying `to_dense` to sparse inputs). 
Since masked sum uses `torch.sparse.sum` then, for the simplicity of masked reductions implementations, its reduction behavior ought to be defined by the behavior of the `torch.sum`. This PR implements the behavioral connection with respect to the directional summation of empty sparse tensors that correspond to all-zero strided tensors.




cc nikitaved pearu cpuhrsch

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33651750).
